codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -411,8 +420,24 @@ private void writeHeadersFrame(
             // We should not re-use ids.
             assert old == null;
 
+            ChannelPromise writeHeadersPromise = ctx.newPromise();
+
+            // Clean up the stream being initialized if writing the headers fails.
+            writeHeadersPromise.addListener(new GenericFutureListener<Future<Void>>() {
+                @Override
+                public void operationComplete(Future<Void> operationFuture) {
+                    if (!operationFuture.isSuccess()) {
```

Comment:
nit: you could reduce the volatile reads that are done internally by the future by using:

```java
Throwable cause = operationFuture.cause();
if (cause != null) {
    frameStreamToInitializeMap.remove(streamId);
    promise.setFailure(operationFuture.cause());
} else {
    promise.setSuccess(operationFuture.getNow());
}
```